### DISCUSSION TITLE
Slingserver on Synology

### DISCUSSION THREAD

**Discussion by donscreen** [GerryDazoo/Slinger#177] [Q&A]
Any chance someone could post a guide to installing the slinger server on a Synology box?   I have Python 3.8 installed on mine and I can SSH into my Synology but I have limited Linux knowledge.  My setup works perfectly on my office Windows desktop machine.   The PIP function seems to be missing in my Python 3.8.

**Comment by GerryDazoo** [comment]
Use   /path_to_python -m pip install flask
What model of Synology box are you using.
Make a separate folder to stick your files in
mkdir sling    for example
You'll need a new sling.service file.
[Unit]
Description=SlingBox Server Service
After=multi-user.target

[Service]
Type=idle
User=gerry
WorkingDirectory=/var/services/homes/gerry/sling
ExecStart=/bin/bash -ce "exec /var/services/homes/gerry/sling/bin/python -u slingbox_server.py >> /volume1/NAS/temp/sling.log 2>&1"
Restart=always

[Install]
WantedBy=multi-user.target

You will need to adjust for your particular setup.

**Reply by donscreen** [reply]
Thanks for the detailed reply Gerry.  My very limited knowledge of Linux is not enough to make this work.  I sincerely appreciate all of the great work you have done.

**Reply by GerryDazoo** [reply]
Did you read the Linux instructions in the readme.txt file. I was assuming you had.

**Reply by donscreen** [reply]
I am running a Synology DS213J.   It is too basic to run Docker.  Docker is not supported by it.  I did read the Linux instructions in the readme.txt but not making any headway with it, mostly due to my lack of Linux knowledge.   I will stick with the working Windows server.  Thank you Gerry!

**Reply by ChrizK-oldGit** [reply]
> Did you read the Linux instructions in the readme.txt file. I was assuming you had.

I have been reading comments about flask on Synology NAS (redit, synology forum). @donscreen also mentions docker.
Would I have to create a docker environment to run perl, or could/should it be installed natively?
(docker will be a whole new learning exercise)

**Reply by f520k510** [reply]
I don't think it is necessary to install python/perl separately if you run it from docker.
In addition setting slinger up in docker is relatively easier than setting it up in ssh.
That assumes your nas support docker of course.

**Reply by ChrizK-oldGit** [reply]
> I don't think it is necessary to install python/perl separately if you run it from docker.

in some ways, I was hoping not to use Docker as it is something I haven't tried before.  That said, everything is an opportunity to learn.
I have a DS918+ and upgrade RAM to 8GB after reading about suggested Docker requirements mid last year...just in case. That could be fortunate :)
Great information from @rbf121 below.  Hopefully get some time over the weekend (must admit, don't get 'Go to Registry' ... I guess it is a Docker thing that may become apparent.

**Reply by GerryDazoo** [reply]
You can install Python3.9 from the package manager. Then "most" of the Linux install instructions apply.

**Comment by rbf121** [comment]
Or run it in docker, no command line needed.

This is on DSM 7:
Install docker from package center.
Open Docker app
Goto Registry
Search for rbf121/slinger
Download
Goto Images
Select rbf121/slinger:latest (whatever version you selected)
Follow on screen prompts
- use the same network as Docker Host
- next
- Advanced Setting is you want to change timezeone (TZ)
- Next
- Add folder
- Select where you want to store your config file (this is persistent storage if you need to deploy a new image later).  I put all my stuff in a docker folder so something like 
- File/Folder: /docker/slinger/config
- Mount Path: /config
- Next
- Done

On first run it will copy all config files and samples into the /config mount path.  
Go to "Containers" in docker app and you should see rbf121-slinger1 running.
Open that and you should see Log, Terminal etc.

Stop the container, go update you config.ini file.  
Start docker.  
Success?

**Reply by GerryDazoo** [reply]
This looks like a good way of doing it. I'll need to give it a go on my Synology box.
Warning: I had a video streaming issue on my NAS. Turns out it was a PC on my network pounding away on the NAS doing a backup. I haven't tried to reproduce it but "niceing" the process might be a good thing to do.

**Reply by f520k510** [reply]
Thanks rbf121.  It's my first time setting up docker on my synology nas.  Much to my surprise, it was very straight forward and I got it to run in less than 10 minutes.  Did you create the registry/template on the docker app in synology package centre?
Thanks again.  Now it is time to toss my raspberry pi 4 as it is purchase specifically for serving as a s slingbox server.

**Reply by rbf121** [reply]
Yes, I created the Dockerfile which has all the installation requirements and initial run steps.  I then published that Docker image on hub.docker.com.  

You can also build your own image using the Dockerfile and push that out to Synology Docker as well without going through docker hub registry, if needed.  That's what I was doing initially but figured I would publish it on docker hub in case anyone else could use it.  It's currently on version 3.0.8 but I see Gerry pushed out a version 3.0.8c with some bugfixes so I should probably update the image with the latest code.

Links:
[Dockerfile](https://github.com/rbf121/Slinger/blob/main/Dockerfile)
[Docker Hub Link](https://hub.docker.com/r/rbf121/slinger)

Hang on to the raspberry pi 4, there are a lot of cool projects you can use it for, check out Retropie if you are interested in retro gaming.  I built a mini arcade using a raspberry pi, spare parts and a 3d printer.

**Reply by DoubleNaught7** [reply]
Hey @rbf121 I'm hoping you can help. I'm running your docker image on my Synology and it's working great. However, I can't seem to figure out how to load a remote file. I always get these two lines in the log:

Reading Custom Remote definition from sample_remote.txt
Error reading remote definition file, using defaults

In my stand-alone Slinger instance on my Mac that I'm trying to replace, I had the files that I used in the same folder as the slingbox_server.py code. Is there a way to map the drives to work properly?

**Reply by rbf121** [reply]
The dockerfile will still copy all of Gerry's sample remote files over to "/usr/src/app/CustomRemotes" but the workdirectory is "/usr/src/app" so you could just reference them like "CustomRemotes/FIOS_remote.txt" for example.  I didn't add them to the startup script to move the files to /config path like the other sample files but you could easily add that if you wanted to in the entrypoint.sh.  Honestly, my use case doesn't require the remote so I don't use anything other than the occasional default remote.  It's possible there could still be an issue with this because I just never used it.

Are you using one of the included CustomRemotes or something else?

If you put a custom remote.txt file (or whatever new remote you want to use) in the same persistent storage area you have the config.ini, then you would reference it in config.ini by using path "/config/remote.txt"

By the way, the docker is a lightweight linux image but it supports bash terminal so you could open up a terminal in synology docker app and navigate through to see the file structure, if needed.

Example:
```
rbf121-slinger3:/usr/src/app/CustomRemotes# ls -ltr                                                                                       
total 1096                                                                                                                                
-rwxr-xr-x    1 root     root        508427 Nov 24 13:16 tivoremotefilesv2.zip                                                            
-rwxr-xr-x    1 root     root         81613 Nov 24 13:16 tivoremotefiles.zip                                                              
-rwxr-xr-x    1 root     root        271938 Nov 24 13:16 skyremotefiles.zip                                                               
-rwxr-xr-x    1 root     root          4201 Nov 24 13:16 shawcable.xfinity.comcastX1.XR5.remote.txt                                       
-rwxr-xr-x    1 root     root          2904 Nov 24 13:16 remote_directv-h24_large_digits.txt                                              
-rwxr-xr-x    1 root     root          2633 Nov 24 13:16 remote_directv-h24.txt                                                           
-rwxr-xr-x    1 root     root          6568 Nov 24 13:16 Tivo.Series2-3.Remote.txt                                                        
-rwxr-xr-x    1 root     root          5850 Nov 24 13:16 Tivo.Premiere.Remote.txt                                                         
-rwxr-xr-x    1 root     root          5705 Nov 24 13:16 Tivo.Mini.Remote.txt                                                             
-rwxr-xr-x    1 root     root          5871 Nov 24 13:16 Tivo.Bolt.Remote.txt                                                             
-rwxr-xr-x    1 root     root         46677 Nov 24 13:16 TIVOremotePF.zip                                                                 
-rwxr-xr-x    1 root     root          6823 Nov 24 13:16 TIVORemote.html                                                                  
-rwxr-xr-x    1 root     root        103674 Nov 24 13:16 ShawBluecurveWesternCanadaHTMLRemote.zip                                         
-rwxr-xr-x    1 root     root          1029 Nov 24 13:16 ProHD_Tuner.txt                                                                  
-rwxr-xr-x    1 root     root          4200 Nov 24 13:16 PACE_RNG200N_Basic.txt                                                           
-rwxr-xr-x    1 root     root          6316 Nov 24 13:16 HDS600rs.txt                                                                     
-rwxr-xr-x    1 root     root          8545 Nov 24 13:16 FIOS_remote.txt                                                                  
-rwxr-xr-x    1 root     root          5584 Nov 24 13:16 BellFibeRemote.txt
```

**Reply by DoubleNaught7** [reply]
Thank you so much for your quick help with this. I've been able to use both methods (adding /config/remote.txt to access my config volume and referencing the CurstomRemotes in the app file) to load .txt remotes. Of course, my preferred remote is TIVORemotePF, which is html based and  included in Gerry's sample remote files as a zip. I have the unzipped .html file and directory in my config drive, and the remote "works" but is missing the css and graphics that make it look great. 

The html file of this remote usually is stored in the app folder along with a tivormotefiles directory that contains the css and graphics. When I use the config.ini include=/config/Remote/TIVORemotePF.html, the remote loads, but it looks terrible and I get the following errors for the missing files:

http://Internal_IP:8080/tivoremotefiles/layout.css
http://Internal_IP:8080/tivoremotefiles/tivoimage.jpg
http://Internal_IP:8080/tivoremotefiles/tvpower.png
http://Internal_IP:8080/tivoremotefiles/tivobutton.png
http://Internal_IP:8080/tivoremotefiles/tivoimage.jpg

Is there any way to change the config file to include that directory or can something be done with the entrypoint.sh script. I'm a bit lost on what to do.

**Reply by rbf121** [reply]
I don't really know anything about the tivoremotefiles but I'm sure there is probably a thread somewhere on github on that.  Do you have the files in the path like this "/config/Remote/TIVORemotePF.html" and "/config/Remote/tivoremotefiles"?

If it requires that the files run in the same directory as the python .py file then you could either try copying the files to /usr/src/app/ so they are set up that way (you will need to do this anytime the docker container is reset or you build a new image yourself with the files.). 

Another option you can try to create a file in /config path caller "slinger.txt", this will override the entrypoint.sh settings.  Copy the python file to /config as well and just have it all run in your /config folder.

/config/slinger.txt
```
python3 -u /config/slingbox_server.py /config/config.ini
```

then have the html file and the tivoremotes file in /config path as well.

**Reply by DoubleNaught7** [reply]
Thanks again @rbf121. I've been trying many things and at the point where you answered my last question, I did create a Remote directory and had the html file and the tivoremotefiles directory located there. I previously had both in config directory.

 I tried your option to use the slinger.txt file and copied the slingbox_server.py into the config directory. I can stream and access the remote, however, the files in the tivoremotes directory do not load and I get the same 404 errors again. My logs are showing that I'm running the version from the config directory:

Version : 3.08f Running on Linux-4.4.59+-x86_64-with pid= 9 /config/slingbox_server.py

The application, remote html file and the tivoremotes directory are all in the config directory, but I get the exact same 404 errors:

http://Internal_IP:8080/tivoremotefiles/layout.css
http://Internal_IP:8080/tivoremotefiles/tivoimage.jpg
http://Internal_IP:8080/tivoremotefiles/tvpower.png
http://Internal_IP:8080/tivoremotefiles/tivobutton.png

Is there anyway to expose the tivoremotes directory with a command or configuration so it can be accessed by a browser?

**Reply by f520k510** [reply]
Hi, I have been trying to tackle the remote issue as well.
If I reference the remote.txt in config.ini with /config/remote.txt, it loads fine.
However, when I replaced remote.txt with skyhdRemote.html, with the jpg file in the folder, skyremotefiles, it failed to load.
I then tried to view the file structure by ssh to nas.  However, I can’t seem to locate the /usr/src/app directory.
I am using ds218+.
Maybe they have different file structure for different models?
If you can tell me where the workdirectory is, it is appreciated.

**Reply by GerryDazoo** [reply]
By default code looks in the folder where slingplayer desktop is for the remote file and as you know you can load the file from anywhere by specifing  the complete path to the remote file, in this case TivoRemote.html.
The issue here is the html code doesn't know where you put the tivoremotefiles so it can download them, You must unzip the tivoremotefiles in the same folder as slingbox_server.  So there is a tivoremmotefiles subfolder.

**Reply by DoubleNaught7** [reply]
I tried @rbf121's suggestion of running a copy of slingbox_server.py from the config folder where my TIVORemote.html file and unzipped tiveoremotes directory reside and I still got the errors.

Since I'm using a TiVo as my source, I've decided to use the Peanut app as my primary remote (it works great, so far) and I'll try to clean up the html file on the remote to remove the files in the tivoremotes and have that as a not so pretty backup.

**Reply by rbf121** [reply]
Try creating another docker volume mapped to /usr/src/app/static

Then putting the tivoremotes static files into the volume should do the trick.  It will create the path /usr/src/app/static/tivoremotes/  which is where the files need to be.

Also, update the html file in your /config/ folder to use the static path. ie "/static/tivoremotefiles/layout.css",  src="/static/tivoremotefiles/tivoimage.jpg", etc.

Just tested it and this seems to work fine.

**Reply by DoubleNaught7** [reply]
YES!! That did it! Of course, I had to complicate things for myself by not following your directions. I initially mapped the docker volume to /static, rather than the /usr/src/app/static that you specified. That led to lots of unsuccessful trial and error until I reread your post and saw what I had done. Thanks so much @rbf121 for figuring this out for me and of course thanks to @GerryDazoo for the slinger project!

**Reply by rbf121** [reply]
Nice.  Gave me a chance to actually try the html version of the remote, it actually looks pretty nice.  I wonder how difficult it would be to adjust it for another remote/cablebox?

**Reply by f520k510** [reply]
> YES!! That did it! Of course, I had to complicate things for myself by not following your directions. I initially mapped the docker volume to /static, rather than the /usr/src/app/static that you specified. That led to lots of unsuccessful trial and error until I reread your post and saw what I had done. Thanks so much @rbf121 for figuring this out for me and of course thanks to @GerryDazoo for the slinger project!

Hi, can you please explain in more details what you did to make it work?
I don't know how to map the docker volume to /usr/app/static


I followed the instructions for slinger to work on nas:

This is on DSM 7:
Install docker from package center.
Open Docker app
Goto Registry
Search for rbf121/slinger
Download
Goto Images
Select rbf121/slinger:latest (whatever version you selected)
Follow on screen prompts

use the same network as Docker Host
next
Advanced Setting is you want to change timezeone (TZ)
Next
Add folder
Select where you want to store your config file (this is persistent storage if you need to deploy a new image later). I put all my stuff in a docker folder so something like
File/Folder: /docker/slinger/config
Mount Path: /config
Next
Done

I cannot make the skyhdRemote.html to show up correctly but a .txt remote would work with no problem.

**Reply by DoubleNaught7** [reply]
@f520k510 Stop your container and go to the Volume tab add a new folder named static located in your slinger folder (at the same level as config) and set the mount point to /usr/src/app/static. Put your directory that your remote files are in in there, mine was tivoremote. Then make the changes to the html remote to use the static path as listed in @rbf121's post above:  ie. "/static/tivoremotefiles/layout.css", src="/static/tivoremotefiles/tivoimage.jpg", etc. Once that's done restart the docker.

**Reply by ChrizK-oldGit** [reply]
> * Advanced Setting is you want to change timezeone (TZ)

Hi @rbf121/everyone , sorry if it is obvious to others, but can you tell me the TZ format for UK (London ...I assume I don't have to specify GMT or BST).
Many thanks

**Reply by ChrizK-oldGit** [reply]
> Stop the container, go update you config.ini file. Start docker. Success?

HI @rbf121 , could you confirm **what I should see in the log if connection is successful**?
My concern is that I have no idea what my Sling Media password is, so I can't retrieve the admin password.  I have changed the config.ini password to be 'admin'.  The log captures Slingbox Found but gives no indicates if the password has been authenticated, 'Waiting for first stream' makes me wonder if it is successful???

I have tried to use VLC, Media > Open Network Stream, using http://192.168.0.4:8080/slingbox and http://172.17.01:8080/slingbox but VLC reports "VLC is unable to open the MRL".  Can you confirm I should be using my NAS IP (192.168.0.4)?  I don't know why the log reports 172.17.0.1 as I would have thought that was unreachable????

I would be grateful if you could tell me the timezone format for the UK.  I set it to 'UK' but given the default was US/Eastern, I suspect this is wrong.

Version : 3.08f Running on Linux-4.4.180+-x86_64-with pid= 15 /usr/src/app/slingbox_server.py
Using config file /config/config.ini
Connection Manager Running on port 8080 with 10 max streams using URL slingbox.
Streamer Running:  10 /config/config.ini SLINGBOX starting up on  port 80808080 8080  8388608
8388608
 * Serving Flask app 'slingbox_server'
 * Debug mode: off
8080 No valid slingbox ip info found in config.ini
Finding Slingbox on local network. My IP Info =  192.168.0.4
Slingbox Found 192.168.0.194 5301 " My Family " FinderID D38.xxxxxxxxxxxxxxxEDxx
Finding Slingbox on local network. My IP Info =  172.17.0.1
8080 Using slingbox at  ('192.168.0.194', 5301)
Streamer:  8080 Waiting for first stream, flushing any IR request

**Reply by GerryDazoo** [reply]
Yes, use the NAS IP. 
http://192.168.0.4:8080/slingbox  should be correct. But I can't say for sure without seeing your config.ini info. Maybe a firewall issue on the NAS?
You should grab the latest server code V4.01.
And give SlingerPlayer a try. VLC is the bane of my existence.  https://github.com/GerryDazoo/SlingerPlayer 

RE:172.17.0.1. Your NAS has mutiple NICs. The server code tries them all to find any accessible Slingboxes. run the "ip address" command to get the list.

When you start a stream you should see something like this:
03/25/2023, 10:58:29.666   Streaming request from ('192.168.117.122', 53420)  /slingbox
STREAM=192.168.117.122:53420:0

Slinginfo  350/500 12 60 192.168.117.13 5001 3072 10 admin
Test Starting Stream for  192.168.117.122:53420
Test Opening Control stream 0x2000 0 0
Connecting... ('192.168.117.13', 5001) Control
Test Selecting Video Source 0
Test Keycodes= [1, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 21, 24, 25, 26, 27, 28, 29, 30, 31, 32, 34, 35, 36, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 62, 63, 64, 66, 67, 68, 69, 81]
VideoParameters: Resolution= 12 FrameRate= 60 VideoBandwidth= 2000 VideoSmoothness= 63 IframeRate= 5 AudioBitRate= 64
Connecting... ('192.168.117.13', 5001) Stream
Test Stream started at 03/25/2023, 10:58:39.34  732 0

**Reply by ChrizK-oldGit** [reply]
Thanks @GerryDazoo , although it was interesting installing Docker and the image from @rbf121, I came to the conclusion it isn't that easy to update files other than those mounted in config (I thought I could mount /usr/src/app and just 'update' but the container wouldn't start).  The information from @f520k510 looks really interesting, so I think I will try this next, allowing me to use the V4.01 ... more learning ahead ...

**Reply by garlandpr** [reply]
> Or run it in docker, no command line needed.
> 
> This is on DSM 7: Install docker from package center. Open Docker app Goto Registry Search for rbf121/slinger Download Goto Images Select rbf121/slinger:latest (whatever version you selected) Follow on screen prompts
> 
> * use the same network as Docker Host
> * next
> * Advanced Setting is you want to change timezeone (TZ)
> * Next
> * Add folder
> * Select where you want to store your config file (this is persistent storage if you need to deploy a new image later).  I put all my stuff in a docker folder so something like
> * File/Folder: /docker/slinger/config
> * Mount Path: /config
> * Next
> * Done
> 
> On first run it will copy all config files and samples into the /config mount path. Go to "Containers" in docker app and you should see rbf121-slinger1 running. Open that and you should see Log, Terminal etc.
> 
> Stop the container, go update you config.ini file. Start docker. Success?

Question, Can I run docker in my home synology  in USA if the slingbox is on another location/Country?
I followed all the info you listed but the log is telling me "can't find a slingbox on network"
Thanks

**Reply by glaypeng78** [reply]
Hi @rbf121 , I was using your docker image on Synology nas and really appreciate your work. However I was facing time out problem that forcing me to re-connect several time before the stream works. I want to try the latest 4.0.1 version of slinger but unfortunately I am lack of experience to modify docker images. Could you update to the 4.0.1 version if not busy or could you tell the exact process to update the file in docker images. So that I can follow the steps to update by myself.

**Comment by Toei79** [comment]
whats the plus for using it on synology nas?  
i have my own but i dont see difference vs linux server that i run on thinkcentre pc, raspberry pi 4 cant handle many stuff in same time?

**Reply by rbf121** [reply]
Synology has an easy to use docker package which I can use to easily view logs or manage remotely.  Plus it's one less server/device to mess around with, ie I have slinger, xteve, plex, tautulli all on same host.

**Reply by Toei79** [reply]
true its best all in one. 

i never able to figure out with docker there . just use basic stuff but i have one pi and thinkcentre like linux servers because that i guess.

**Comment by f520k510** [comment]
just installed slinger server on synology with ssh and thought I would post it here for the record.
First I upgrdaed dsm to 7.0 so that I can put sling.service in /etc/systemd/system/ and use systemctl to enable it.
This is because starting dsm 7.0 synology uses systemd and systemctl to manage service.
According to the change log starting 7.0 python3 has become part of the system and don't need to install from package center.
Then I enabled ssh in control panel.  here:  control panel->terminal & snmp
Then I created an admin user called "slingbox"
I made a folder inside /homes/slingbox with file explorer and named it slinger.  Copied sling.service, slingbox_server.py, remote.txt, and config.ini to the folder slinger.  Made sure config.ini is working.
Then, I ssh to synology with user slingbox,  
installed pip with the command:
sudo python3 -m ensurepip
upgraded pip with the command:
sudo python3 m pip install --upgrade pip
installed flask module with:
sudo python3 -m pip install flask
edit sling.service with vi and paste the following contents:
vi sling.service

"
[Unit]
Description=SlingBox Server Service
After=multi-user.target

[Service]
Type=idle
User=slingbox
WorkingDirectory=/var/services/homes/slingbox/slinger
ExecStart=/bin/python3  slingbox_server.py
Restart=always

[Install]
WantedBy=multi-user.target
"

change mode of the file slingbox_server.py with this:
sudo chmod +x /var/services/homes/slingbox/slinger/slingbox_server.py
reloaded daemon, enabled the service, and started it:
sudo systemctl daemon-reload
sudo systemctl enable sling.service
sudo systemctl start sling.service


If all goes well, check the status of the service:
sudo systemctl status sling.service

It should say something like this:

● sling.service - SlingBox Server Service
   Loaded: loaded (/etc/systemd/system/sling.service; enabled; vendor preset: disabled)
   Active: active (running) since Sat 2023-03-25 22:26:18 CST; 323ms ago
 Main PID: 15955 ((python3))
   CGroup: /system.slice/sling.service
           └─15955 (python3)

Mar 25 22:26:18 nas systemd[1]: Started SlingBox Server Service.
Mar 25 22:26:18 nas systemd[1]: Starting SlingBox Server Service...

Hope this helps.

**Reply by ChrizK-oldGit** [reply]
@f520k510 thank you for the instructions ...I will be trying this next.  Are the instructions based on a published article for another installation?  If so, I would appreciate a link as it may answer my question regarding 'operability/usability', that being ...
Running a virtual environment will be new to me.  My understanding is that slingbox_server would run as a background service. Is there a way to view the console, ie the stdout generated by slingbox_server.py?

If this is not possible, it would be great if the stdout could be directed to a log file (which could be tailed).
I guess it would need something like logrotate for management ...of course, this may not be necessary if  a virtual environment allows you to view a console.
Can Synology Log Center be used to view and manage 'third party' logs?

**Reply by f520k510** [reply]
It is not based on a published article.  It is based on my own installation experience.
I basically google information when I needed to and trial and error when I had to.
I also followed the instruction that is on the code tab of slinger on github which Gerry gave us.
I found the information that I needed regarding virtual environment here:
https://docs.python.org/3/library/venv.html
I don't think it is absolutely necessary to run slinger in a virtual environment.
In fact I run it in a virtual environment on my friend's nas, but I am not running it on my own nas.
I will update my instruction to remove the virtual environment part now. 

Anyway, if you run into any issues, just let me know.  I am happy to help you if I can.

**Reply by ChrizK-oldGit** [reply]
EDIT *** I am going to leave this comment as it may be relevant for others ***
I believe the instructions provided expect the user home service to be enabled.
Control Panel > User and Group > Advanced > User Home, Enable user home service
This will create a user home directory under /volume(x)/homes.  I do not have this enabled, and so I do not have /homes/slingbox as expected when the user is created in an earlier step.  It also shows I have been confused with the configuration showing a similar path, and answers that the slinger files are not stored within the services path. 
Given that I do not have /volume(x)/homes, I am going to proceed with the files under /opt/slinger (Add-on Applications)
= = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = =

Hi @f520k510 I am returning to my NAS for installation, but have failed very early on in your instructions.
Do the slinger files need to reside under /var/services/homes/?
Your instructions say to make 'a folder inside /homes/slingbox with file explorer and name it slinger'
This is where I am failing as Filer has no shared folder for /homes, and if I try to navigate to it, I am told that the directory does not exist.

If I SSH (sudo su) I find that homes is a link, although I can't follow it..
Can you tell me where I am going wrong?
Many thanks

root@mynas:/var/services# **ls -l homes** 
lrwxrwxrwx 1 root root 24 Sep 12  2021 homes -> **/volume1/@fake_home_link**
root@mynas:/var/services# **cd homes** 
-ash: cd: homes: No such file or directory
root@mynas:/var/services# **realpath homes** 
/volume1/@fake_home_link
root@mynas:/var/services# **cd /volume1/@fake_home_link**
-ash: cd: /volume1/@fake_home_link: No such file or directory

**Reply by f520k510** [reply]
Maybe I didn't make it clear, but I meant you should create a user called slingbox.  Then dsm should create a folder called slingbox inside /homes.  You then create a folder called slinger.  Then if you ssh to the gas, you should see that the present working directory is /var/services/homes/slingbox/

Let me know if you have any further questions.

**Reply by ChrizK-oldGit** [reply]
> Maybe I didn't make it clear, but I meant you should create a user called slingbox.

Thank you.  Really appreciate your instructions.  I have updated my initial comment as I believe there is a pre-requisite for User Home Service.  My DSM has been upgraded over the years.  Maybe a fresh install of DSM 7 would have enabled this by default.

I think you may have missed a step, before enabling the service, ie to copy the sling.service file to /etc/systemd/system
I assume this is done as root to copy with user and group ownership.

```
root@mynas:/opt/slinger# cp sling.service /etc/systemd/system
root@mynas:/etc/systemd/system# ls -l sling.service 
-rwx------ 1 root root 225 Apr  2 12:19 sling.service
root@mynas:/etc/systemd/system# sudo systemctl enable sling.service
Created symlink from /etc/systemd/system/multi-user.target.wants/sling.service to /etc/systemd/system/sling.service.

root@mynas:/etc/systemd/system# sudo systemctl start sling.service
root@mynas:/etc/systemd/system# sudo systemctl status sling.service
● sling.service - SlingBox Server Service
   Loaded: loaded (/etc/systemd/system/sling.service; enabled; vendor preset: disabled)
   Active: active (running) since Sun 2023-04-02 12:28:05 IST; 18s ago
 Main PID: 22532 (python3)
   Memory: 6.3M
   CGroup: /system.slice/sling.service
           └─22532 /bin/python3 slingbox_server.py

Apr 02 12:28:05 mynas systemd[1]: Started SlingBox Server Service.
Apr 02 12:28:05 mynas systemd[1]: Starting SlingBox Server Service...

```
@GerryDazoo , if I have started slingbox_server correctly using this method, should I get a log file?

```
root@mynas:/etc/systemd/system# cd /tmp
root@mynas:/tmp# ls sling.log
**ls: cannot access 'sling.log': No such file or directory**

root@mynas:/# find . -name  sling.log
root@mynas:/#
```

**Reply by ChrizK-oldGit** [reply]
I noted that the suggested sling.service file does not include the log lines that are in the README, however, adding these appears to be causing me a problem.

Also, I am fairly sure that although the previous status (no log) indicated that the server was running, it was not.  At least, not with the expected port.  **netstat -na | grep 8080** returns no result :(

```
root@my_nas:/etc/systemd/system# sudo systemctl status -l sling.service
● sling.service - SlingBox Server Service
   Loaded: loaded (/etc/systemd/system/sling.service; enabled; vendor preset: disabled)
   Active: active (running) since Sun 2023-04-02 12:52:57 IST; 4min 18s ago
 Main PID: 30696 (python3)
   CGroup: /system.slice/sling.service
           └─30696 /bin/python3 slingbox_server.py

Apr 02 12:52:57 my_nas systemd[1]: Started SlingBox Server Service.
Apr 02 12:52:57 my_nas systemd[1]: Starting SlingBox Server Service...
Apr 02 12:53:10 my_nas systemd[1]: [/etc/systemd/system/sling.service:9] Failed to parse output specifier, ignoring: append:/tmp/sling.log
Apr 02 12:53:10 my_nas systemd[1]: [/etc/systemd/system/sling.service:10] Failed to parse output specifier, ignoring: append:/tmp/sling.log
Apr 02 12:53:13 my_nas systemd[1]: Started SlingBox Server Service.
```

**Reply by f520k510** [reply]
Adding the log file in sling.service did cause me a problem, too.  That's why I removed it.
If you need the log for debug, I suggest you run slingbox_server.py from ssh command with python3 slingbox_server.py, instead of trying to create a log file or debug.  After you make sure everything is working, you can then create and enable the sling.service.

**Reply by ChrizK-oldGit** [reply]
EDIT *** Ignore the question, but keeping for reference ***
I had left the service running ...sorry

NB The reason the service was failing, I had to rename config.ini as I transferred it as config.ini.synology due to all my different attempts and investigations.
Also, while stopping (Ctrl-C) and starting slingbox_server.py from command line I had a few instances of **OSError: [Errno 98] Address already in use**
It appears that the process used to listen on the port doesn't always close.  I had to kill the python3 process to start the server once more.

sh-4.4$ sudo netstat -nlp | grep 8080
tcp        0      0 0.0.0.0:8080           0.0.0.0:*               LISTEN      6095/python3      
sudo kill -9 6095

```
Exception in thread Thread-1:
Traceback (most recent call last):
  File "/usr/lib/python3.8/threading.py", line 932, in _bootstrap_inner
8080 Checking for slingbox at ('192.168.0.15', 5301) 0
    self.run()
  File "/usr/lib/python3.8/threading.py", line 870, in run
    self._target(*self._args, **self._kwargs)
  File "slingbox_server.py", line 1133, in ConnectionManager
    ConnectionManagerSocket.bind(server_address)
OSError: [Errno 98] Address already in use
```
= = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = =

Thanks, that was a good suggestion.
Does anybody know what **Address already in use** is referring to?

```
Version : 4.01 Running on Linux-4.4.180+-x86_64-with-glibc2.2.5 pid= 25527 slingbox_server.py
Using config file config.ini
Connection Manager Running on port 8080 with 10 max streams using URL slingbox.
Building page for Slingbox
Using built in default remote page definition.
Broadcast Responder Running
Starting Streamer Thread for Slingbox
ERROR While handling GETPORTNUMBERS request [Errno 98] Address already in use
Streamer Running:  10 config.ini SLINGBOX 8080 8080 8388608
starting up on  port 8080 8388608
8080 Checking for slingbox at ('192.168.0.15', 5301) 0
8080 ('192.168.0.15', 5301) OK
8080 Using slingbox at  ('192.168.0.15', 5301)
Streamer:  8080 Waiting for first stream, flushing any IR requests that arrive while not connected to slingbox
ERROR While handling GETPORTNUMBERS request [Errno 98] Address already in use
ERROR While handling GETPORTNUMBERS request [Errno 98] Address already in use
ERROR While handling GETPORTNUMBERS request [Errno 98] Address already in use
ERROR While handling GETPORTNUMBERS request [Errno 98] Address already in use
ERROR While handling GETPORTNUMBERS request [Errno 98] Address already in use
```

**Reply by GerryDazoo** [reply]
It means you've got multiple copies of the server code running.

**Comment by GerryDazoo** [comment]
The NAS uses a different version of systemd than the version used on the Pi. So the one supplied will not work for specifying the log file output. This is what I have on my NAS.

`[Unit]
Description=SlingBox Server Service
After=multi-user.target

[Service]
Type=idle
User=gerry
WorkingDirectory=/var/services/homes/gerry/sling
ExecStart=/bin/bash -ce "exec /var/services/homes/gerry/sling/bin/python -u slingbox_server.py >> /volume1/NAS/temp/sling.log 2>&1"
Restart=always

[Install]
WantedBy=multi-user.target
`
The log file is sent to a shared folder on the NAS so you can review it without having to login via SSH.

**Comment by glaypeng78** [comment]
> Or run it in docker, no command line needed.
> 
> 
> 
> This is on DSM 7:
> 
> Install docker from package center.
> 
> Open Docker app
> 
> Goto Registry
> 
> Search for rbf121/slinger
> 
> Download
> 
> Goto Images
> 
> Select rbf121/slinger:latest (whatever version you selected)
> 
> Follow on screen prompts
> 
> - use the same network as Docker Host
> 
> - next
> 
> - Advanced Setting is you want to change timezeone (TZ)
> 
> - Next
> 
> - Add folder
> 
> - Select where you want to store your config file (this is persistent storage if you need to deploy a new image later).  I put all my stuff in a docker folder so something like 
> 
> - File/Folder: /docker/slinger/config
> 
> - Mount Path: /config
> 
> - Next
> 
> - Done
> 
> 
> 
> On first run it will copy all config files and samples into the /config mount path.  
> 
> Go to "Containers" in docker app and you should see rbf121-slinger1 running.
> 
> Open that and you should see Log, Terminal etc.
> 
> 
> 
> Stop the container, go update you config.ini file.  
> 
> Start docker.  
> 
> Success?

Hi, I was using your docker image on Synology nas and really appreciate your work. However I was facing time out problem that forcing me to re-connect several time before the stream works. I want to try the latest 4.0.1 version of slinger but unfortunately I am lack of experience to modify docker images. Could you update to the 4.0.1 version if not busy or could you tell the exact process to update the file in docker images. So that I can follow the steps to update by myself.
